### PR TITLE
`rustc_borrowck` cleanups

### PR DESCRIPTION
**Pull request by nnethercote**
A bunch of cleanups I made while reading over this crate.

r? @lqd

### CODE REVIEW

File: compiler/rustc_borrowck/src/lib.rs [hunk 1]
```diff
@@ -584,6 +566,49 @@ struct MirBorrowckCtxt<'a, 'infcx, 'tcx> {
     move_errors: Vec<MoveError<'tcx>>,
 }
 
+impl<'a, 'infcx, 'tcx> MirBorrowckCtxt<'a, 'infcx, 'tcx> {
+    fn new(
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/lib.rs:569]: I don't think this is cleaner, tbh. I think initializing `MirBorrowckCtxt` with the struct initialization context was totally fine, and now we're just roundtripping through a `new` method that needs to stay in sync with the billion fields that the struct has.

File: compiler/rustc_borrowck/src/type_check/free_region_relations.rs [hunk 2]
```diff
@@ -57,7 +57,7 @@ pub(crate) fn create<'tcx>(
     universal_regions: Rc<UniversalRegions<'tcx>>,
     constraints: &mut MirTypeckRegionConstraints<'tcx>,
 ) -> CreateResult<'tcx> {
-    UniversalRegionRelationsBuilder {
+    CreateResultBuilder {
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/type_check/free_region_relations.rs:60]: This name is really objectively worse. The "`CreateResult`" is, afaict, the "**result** of **creat**ing the universal relations". This basically removes all of the *salient* information about what the struct is used for. I think you should revert this change.

File: compiler/rustc_borrowck/src/diagnostics/region_name.rs [hunk 3]
```diff
@@ -75,83 +75,73 @@ pub(crate) enum RegionNameHighlight {
 
 impl RegionName {
     pub(crate) fn was_named(&self) -> bool {
+        use RegionNameSource::*;
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/diagnostics/region_name.rs:78]: I tend to think that these decrease the readability of these functions. We're really not savingany keystrokes, and all it does is conceptually separate useful information. I'd prefer if all of these `use EnumKind::*;` changes are dropped.

File: compiler/rustc_borrowck/src/type_check/free_region_relations.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_borrowck/src/type_check/free_region_relations.rs:60]: Did you look at how `CreateResult` is being used? It's immediately destructured into its components -- so what *it's* called is not really that meaningful; however, if you care to rename it, I really think it should get a name that is *more* specific, not *less*. 

The main purpose of this `UniversalRegionRelationsBuilder` is to build the `UniversalRegionRelations` are used throughout borrowck, and the only reason that other data is strung along is because it's convenient.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_borrowck/src/lib.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_borrowck/src/lib.rs:569]: Ok, I'll remove it.

File: compiler/rustc_borrowck/src/diagnostics/region_name.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_borrowck/src/diagnostics/region_name.rs:78]: I think there is a spectrum of these. The more compelling ones are when the enum name is long, and there are lots of repetitions of the enum name on adjacent lines, and removing the enum name lets some multi-line patterns collapse into a single line. The `ConstraintCategory` one is the most compelling in this PR. But I can also see I'm unlikely to convince you, so I'll drop these commits.

File: compiler/rustc_borrowck/src/type_check/free_region_relations.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_borrowck/src/type_check/free_region_relations.rs:60]: An idea would be to call it `UniversalRegionRelationsBuilderResult`, though that's a bit long.

Comment (nnethercote) [reply compiler/rustc_borrowck/src/type_check/free_region_relations.rs]: Yeah, I did. I don't like having a type called `FooBuilder` that produces a type not called `Foo`. A deeper clean up is required, I'll do that in a follow-up, and I've removed this commit.

**Comment by nnethercote** [comment]
I've removed the commits you didn't like, should be good to go now.

**Comment by voidc** [comment]
Reducing the visibility of items in this crate might break tools using the `get_body_with_borrowck_facts` API. In #111840, some types and methods were deliberately made public, so that they could be used externally.

**Comment by nnethercote** [comment]
> visibility of items 

I checked. Nothing in the `consumers.rs` file was changed. I thought that file contained the entire public API, but I found there are a few items outside that file marked with a comment as being in the API. I have reverted the visibility changes to those items. Should be good to go now, thanks.

**Comment by lqd** [comment]
as they’ve already started reviewing, r? compiler-errors

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r=me after rebasing
